src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs

```diff
@@ -128,31 +159,122 @@ public IEnumerable<ReplicationBatchItem> GetTombstonesByBucketFrom(DocumentsOper
     {
-        using (GetBucketByteString(context.Allocator, bucket, out var buffer))
-        using (Slice.External(context.Allocator, buffer, buffer.Length, out var prefix))
+        long numOfDeletions = 0;
+
+        MarkTombstonesAsArtificial(context, bucket);
+
+        foreach (var document in GetDocumentsByBucketFrom(context, bucket, 0))
         {
-            return table.DeleteForwardFrom(dynamicIndex, prefix, startsWith: true, MaxDocumentsToDeleteInBucket, shouldAbort: holder =>
-            {
-                var changeVector = TableValueToChangeVector(context, changeVectorPosition, ref holder.Reader);
-                var status = ChangeVectorUtils.GetConflictStatus(changeVector, upTo);
-                return status != ConflictStatus.AlreadyMerged;
-            });
+            if (numOfDeletions > MaxDocumentsToDeleteInBucket)
+                break;
+
+            var docCv = context.GetChangeVector(document.ChangeVector);
+            if (ChangeVectorUtils.GetConflictStatus(docCv, upTo) != ConflictStatus.AlreadyMerged)
+                break;
+
+            // check change vectors of all document extensions
+            if (HasDocumentExtensionWithGreaterChangeVector(context, document.LowerId, upTo)) 
+                break;
+            
+            Delete(context, document.Id, document.Flags | DocumentFlags.Artificial);
+
+            // delete revisions for document
+            RevisionsStorage.DeleteRevisionsFor(context, document.Id, artificial: true);
+
+            numOfDeletions++;
         }
+
+        return numOfDeletions;
     }
 
-    public static long DeleteItemsByBucketForDocuments(DocumentsOperationContext context, CollectionName collection, int bucket, ChangeVector upTo)
+    private void MarkTombstonesAsArtificial(DocumentsOperationContext context, int bucket)
     {
-        var table = context.Transaction.InnerTransaction.OpenTable(DocsSchema, collection.GetTableName(CollectionTableType.Documents));
-        return DeleteItemsByBucket(context, table, DocsSchema.DynamicKeyIndexes[CollectionDocsBucketAndEtagSlice], (int)DocumentsTable.ChangeVector, bucket, upTo);
+        var readTable = new Table(TombstonesSchema, context.Transaction.InnerTransaction);
+
+        foreach (var result in GetItemsByBucket(context.Allocator, readTable, TombstonesSchema.DynamicKeyIndexes[TombstonesBucketAndEtagSlice], bucket, etag: 0))
+        {
+            var tombstone = TableValueToTombstone(context, ref result.Result.Reader);
+            if (tombstone.Flags.Contain(DocumentFlags.Artificial))
+                continue;
+
+            var collection = TableValueToId(context, (int)TombstoneTable.Collection, ref result.Result.Reader);
+            if (_collectionsCache.TryGetValue(collection, out var collectionName) == false)
+            {
+                collectionName = new CollectionName(collection);
+                _collectionsCache.TryAdd(collection, collectionName);
```

Comment:
shouldn't this be like this (copied from DocumentsStorage):

```
                // Add to cache ONLY if the transaction was committed.
                // this would prevent NREs next time a PUT is run,since if a transaction
                // is not committed, DocsSchema and TombstonesSchema will not be actually created..
                // has to happen after the commit, but while we are holding the write tx lock
                context.Transaction.InnerTransaction.LowLevelTransaction.BeforeCommitFinalization += _ =>
                {
                    var collectionNames = new Dictionary<string, CollectionName>(_collectionsCache, StringComparer.OrdinalIgnoreCase)
                    {
                        [name.Name] = name
                    };
                    _collectionsCache = collectionNames;
                };
```

Reply:
done